### PR TITLE
Add inline docs for multiple issues

### LINKED ISSUE #62: Allow abilities to be added to the "XCOM" pool when they exist in the soldiers randomized ability pool
Note: This should only add those abilities that hasn't been populated in the ability tree.

### LINKED ISSUE #384: Forced Appearance Issues
The `bForcedAppearance` Character Template flag relies on `XComHumanPawn::UpdateAppearance()`, which only updates the non-blank elements from the forced appearance. There's two issues however - one it was never updated to account for the new AC and WotC deco slots. Two, since it can't make a slot empty, means that Central for example may pick up deco slots in the generator! My solution would be for XGCharacterGenerator to leave the deco slots empty if bForcedAppearance is set.

### LINKED ISSUE #393: Issue #393 - Add Event Hook to GetClipSize
Add an event hook in XComGameState_Item to allow effects to dynamically modify weapon ammo count.

### LINKED ISSUE #409: Allow mods to add parallel soldier abilities, such as officer abilities, and have them displayed
Much of XCOM 2's UI uses `XComGameState_Unit.GetEarnedSoldierAbilities()` to gain access to a soldier's current abilities, but it's hard-coded to only include class and AWC abilities. The officer ability tree is a parallel set of abilities, so soldiers' officer abilities aren't displayed in the main game.

The highlander should allow mods to add their own abilities. This would enable other systems, like allowing normal classes to also have psi powers, just as in XCOM 1.

### LINKED ISSUE #448: Allow mods to disable reinforcement flares and other RNF visualisations
The visuals that show where reinforcements will spawn can make handling them too easy for players. So allow mods to disable those visuals.

### LINKED ISSUE #475: Allow sources of disorient to affect disoriented units
In vanilla, it's impossible to break the overwatch of a unit that's disoriented with a flashbang because the disorient effect can't be reapplied to disoriented units.

### LINKED ISSUE #476: Make AbilityRequiresTargetingActivation()/IsTargetingMethodActivated() use a configurable list
(Controller Support)

UITacticalHUD_AbilityContainer.AbilityRequiresTargetingActivation does this:
```
if (ClassIsChildOf(AbilityState.GetMyTemplate().TargetingMethod, class'X2TargetingMethod_Grenade') ||
	ClassIsChildOf(AbilityState.GetMyTemplate().TargetingMethod, class'X2TargetingMethod_Cone') ||
	ClassIsChildOf(AbilityState.GetMyTemplate().TargetingMethod, class'X2TargetingMethod_Line') ||
	ClassIsChildOf(AbilityState.GetMyTemplate().TargetingMethod, class'X2TargetingMethod_BlasterLauncher'))
{
	return true;
}
```
`IsTargetingMethodActivated()` does similar. These functions indicates that after selecting the ability, you have to press A again to aim the ability, which allows the user to rotate the camera with the buttons that would normally cycle the ability selection.
However, this skips some TargetingMethods -- RocketLauncher, for example is a Grenade, but grappling and BlazingPinions are not covered by the list. Making this list configurable would allow a more consistent behaviour.

### LINKED ISSUE #500: Provide an option to disable changing patrol zones
Even if you disable the intercept behaviour, the base game will still adjust patrol zones based on changes to the line of play. There should be an option to disable these "dynamic encounter zones" as tracktwo puts them.

### LINKED ISSUE #507: Allow mods to override patrol behaviour
`XComGameState_AIGroup.ShouldMoveToIntercept()` controls patrol behaviour for pods, but some mods (read LW2) may want to manage patrol routes and behaviour themselves. There should be an event that allows mods to disable/bypass the default patrolling logic.

### LINKED ISSUE #534: Add an event notification for when psi training completes
Should go at the end of `XComGameState_HeadquartersProjectPsiTraining.OnProjectCompleted()`.

### LINKED ISSUE #561: Add event in X2SitRepTemplate::MeetsRequirements
This will allow mods to run additional checks

### LINKED ISSUE #564: Allow mods to prevent hack interactions with objects
Add event to `XComGameState_InteractiveObject::CanInteractHack`

### LINKED ISSUE #617: Rocket launchers can fire through blocking tiles
The `X2TargetingMethod_RocketLauncher` class has a bug where if the shooter of the rocket can step out, they can shoot the rocket through walls and other obstructions. This is because the code performs a ray trace between the unit's tile and its step out tile. The target tile isn't involved!

### LINKED ISSUE #624: Allow mods to override respec times for soldiers in GTS
An event hook should allow mods to override the default respec time with a bit more flexibility, for example by incorporating the rank of the soldier in the respect time.

### LINKED ISSUE #631: Allow mods to determine whether to display promotion icon
This is primarily to allow mods to force the display of the promotion icon for Psi Operatives, but it could be used for any changed behaviour. It's required for Psi Operatives so that folks know when they can put an operative back in the tubes for training.

### LINKED ISSUE #633: Allow mods to control which research techs can be inspired
The highlander should fire an event that allows mods to determine whether a tech can be researched or not. This provides them an important lever for balancing the strategy game. It's particularly important for repeatable research, which the inspiration mechanic doesn't seem to handle very well (the techs remain inspired even after the first inspired research of them is complete).

### LINKED ISSUE #695: Prevent CAs from being marked as "rookie CA"
The game has a concept of "rookie covert actions":

> covert actions which do not have veteran soldier requirements, so can always be started

Mods can introduce CAs which do not have rank requirements, but have very specific spawn requirements, which would be ignored by current system. 

I propose adding `X2CovertActionTemplate::bCanNeverBeRookie` to allow mods to control this behaviour (`X2CovertActionTemplate` is not native). The value will be checked in `XComGameState_HeadquartersResistance::FindRookieCovertActions`

### LINKED ISSUE #701: UIArmory_loadoutItem cannot handle tutorial medikit replacements
https://github.com/X2CommunityCore/X2WOTCCommunityHighlander/blob/f3e185762061ab1c399360478d17bb450adcb8af/X2WOTCCommunityHighlander/Src/XComGame/Classes/UIArmory_LoadoutItem.uc#L73

https://github.com/X2CommunityCore/X2WOTCCommunityHighlander/blob/f3e185762061ab1c399360478d17bb450adcb8af/X2WOTCCommunityHighlander/Src/XComGame/Classes/UIArmory_Loadout.uc#L1206-L1207

This affects A Better Barracks.

### PR DESCRIPTION
**Pull request by Iridar**
Closes #62
Closes #617
Closes #701
Closes #695
Closes #633
Closes #631
Closes #624
Closes #564
Closes #561
Closes #534
Closes #507
Closes #500
Closes #384
Closes #393
Closes #476
Closes #475
Closes #448
Closes #409

These issues have already been implemented, this PR just adds docs for them.

### CODE REVIEW

**Comment by Iridar** [comment]
Should this PR have a separate commit for docs for each issue, or should I squash it?

File: X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Item.uc [hunk 1]
```diff
@@ -935,8 +935,21 @@ simulated function int GetClipSize()
...
+	/// The `OverrideClipSize` event allows mods to override Clip Size 
+	/// of a weapon after it has been modified by weapon upgrades and/or loaded ammo.
+	///
+	///	This can help make a passive ability that modifies Clip Size of the soldier's weapon,
+	/// or to explicitly disallow a specific weapon benefitting from effects that modify Clip Size 
+	/// by resetting the clip size value to the clip size value stored in the weapon template.
+	///    
+	///```event
+	///EventID: OverrideClipSize,
+	///EventData: [out int iClipSize],
```
Comment (robojumper) [inline X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Item.uc:949]:
```suggestion
	///EventData: [inout int iClipSize],
```

File: X2WOTCCommunityHighlander/Src/XComGame/Classes/X2SitRepTemplate.uc [hunk 2]
```diff
@@ -118,6 +118,17 @@ function bool MeetsRequirements(XComGameState_MissionSite MissionState)
...
 
 // Start issue #561
+/// HL-Docs: feature:SitRepCheckAdditionalRequirements; issue:561; tags:strategy
+/// The `SitRepCheckAdditionalRequirements` event allows mods to subject sitreps
+/// to additional checks for specific missions, and then either allow the sitrep
+/// to be added to the particular mission (default behavior), or forbid it.
+///    
+///```event
+///EventID: SitRepCheckAdditionalRequirements,
+///EventData: [out bool bMeetsRequirements, in XComGameState_MissionSite MissionState],
```
Comment (robojumper) [inline X2WOTCCommunityHighlander/Src/XComGame/Classes/X2SitRepTemplate.uc:128]:
```suggestion
///EventData: [inout bool bMeetsRequirements, in XComGameState_MissionSite MissionState],
```

File: X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_HeadquartersXCom.uc [hunk 3]
```diff
@@ -6944,19 +6944,18 @@ function CheckForInspiredTechs(XComGameState NewGameState)
...
-//
+/// HL-Docs: feature:CanTechBeInspired; issue:633; tags:strategy
+/// The `CanTechBeInspired` event allows mods to forbid a tech from being inspired. 
+/// This provides an important lever for balancing the strategy game. 
+/// It's particularly important for repeatable techs, which the inspiration mechanic doesn't seem to handle very well 
+/// (the techs remain inspired even after the first inspired research of them is complete).
+///
+///```event
+///EventID: CanTechBeInspired,
+///EventData: [out bool bCanTechBeInspired],
```
Comment (robojumper) [inline X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_HeadquartersXCom.uc:6955]:
```suggestion
///EventData: [inout bool bCanTechBeInspired],
```

File: X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_InteractiveObject.uc [hunk 4]
```diff
@@ -522,6 +522,15 @@ event bool CanInteractHack(XComGameState_Unit Unit)
 }
 
 // Start issue #564
+/// HL-Docs: feature:AllowInteractHack; issue:564; tags:tactical
+/// The `AllowInteractHack` event allows mods to forbid an interactive object from being hackable by a unit. 
+///	
+///```event
+///EventID: AllowInteractHack,
+///EventData: [out bool bAllowInteractHack, in XComGameState_Unit HackingUnit],
```
Comment (robojumper) [inline X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_InteractiveObject.uc:530]:
```suggestion
///EventData: [inout bool bAllowInteractHack, in XComGameState_Unit HackingUnit],
```

File: X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Unit.uc [hunk 5]
```diff
@@ -13209,6 +13209,19 @@ function SCATProgression GetSCATProgressionForAbility(name AbilityName)
...
+	/// HL-Docs: feature:OverrideShowPromoteIcon; issue:631; tags:strategy
+	/// The `OverrideShowPromoteIcon` event allows mods to determine 
+	/// whether a promotion icon for a particular soldier should be displayed or not.
+	/// This can be relevant for mods that add their own promotion mechanics for soldiers,
+	/// e.g. a psionic class that has to go on a few missions before they can be stuck into
+	/// a Psi Lab to get their promotion.
+	///
+	///```event
+	///EventID: OverrideShowPromoteIcon,
+	///EventData: [out bool bShowPromotionIcon],
```
Comment (robojumper) [inline X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Unit.uc:13221]:
```suggestion
	///EventData: [inout bool bShowPromotionIcon],
```

**Review by robojumper — CHANGES_REQUESTED** [review]
One commit per issue is fine.

I'm now a bit unsure about our handling of `inout` vs `out`; I'd say it needs to be `inout` when the default value in the tuple potentially has a different value than the data type's default value just to make it so that the auto-generated event listener preserves the default value.

Like, this should be avoided (from the github-actions comment):

```
	local int iClipSize;

	ItemState = XComGameState_Item(EventSource);
	Tuple = XComLWTuple(EventData);

	// Your code here

	Tuple.Data[0].i = iClipSize;
```

because an unmodified listener will now set the clip size to 0 :)

File: X2WOTCCommunityHighlander/Src/XComGame/Classes/XComHumanPawn.uc [hunk 6]
```diff
@@ -2431,6 +2433,8 @@ static simulated function UpdateAppearance( out TAppearance mAppearance, const o
 		mAppearance.nmVoice = kAppearance.nmVoice;
 	}
 	// Start Issue #384
+	/// HL-Docs: ref:Bugfixes; issue:384
+	/// Allow this function to work with new deco slots added with Anarchy Children and WOTC.
```
Comment (robojumper) [inline X2WOTCCommunityHighlander/Src/XComGame/Classes/XComHumanPawn.uc:2437]:
```suggestion
	/// Allow `bForcedAppearance` to work with new deco slots added with Anarchy Children and WOTC.
```

File: X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_AIGroup.uc [hunk 7]
```diff
@@ -388,14 +388,14 @@ function bool ShouldMoveToIntercept(out Vector TargetInterceptLocation, XComGame
...
-		//
-		// Allow mods to override the LoP anchor point for the encounter zone. By
-		// default, the encounter zone adjusts to the current location of the XCOM
-		// squad.
-		//
-		// As an example, a mod could use the XCOM's spawn location instead so that
-		// the encounter zones remain the same regardless of where the XCOM squad is
-		// currently.
+		/// HL-Docs: ref:OverrideEncounterZoneAnchorPoint
+		/// Allow mods to override the LoP anchor point for the encounter zone. By
```
Comment (robojumper) [inline X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_AIGroup.uc:392]:
```suggestion
		/// This allows mods to override the LoP anchor point for the encounter zone. By
```

**Review by robojumper — APPROVED** [review]
Only small style nits now.

File: X2WOTCCommunityHighlander/Src/XComGame/Classes/UIArmory_LoadoutItem.uc [hunk 8]
```diff
@@ -70,6 +70,8 @@ simulated function UIArmory_LoadoutItem InitLoadoutItem(XComGameState_Item Item,
 		SetDisabled(true, class'UIUtilities_Text'.static.GetColoredText(InitDisabledReason, eUIState_Bad));
 	}
 
+	/// HL-Docs: ref:Bugfixes; issue:701
+	///	Allows armory UI to highlight the item the player needs to build during the tutorial even if it's not a medkit.
```
[the remark this answers is not in the document]
Comment (Iridar) [reply X2WOTCCommunityHighlander/Src/XComGame/Classes/UIArmory_LoadoutItem.uc]: Pulled the rug from under this suggestion by committing previous suggestions, had to do this one locally first.

**Comment by Iridar** [comment]
> One commit per issue is fine.

Okay, I'll squash only the "fix" types of commits then.

**Comment by Iridar** [comment]
All done. These are all the issues I intend to document for now, other open issues with need-docs label put me too far out of my element.

**Comment by Iridar** [comment]
@pledbrook 
@Xymanek 
@MrNiceUK 
@shiremct
This PR contains docs for Issues resolved by you, please take a look to make sure the docs are correct, as my understanding in some of the more niche issues is less than perfect.

I would also encourage you to file docs for your Issues not covered by this PR yourselves; I did not cover them because they are too far out of my specialty.

**Pull request merged**